### ISSUE TITLE
Events

### ISSUE THREAD

**Issue by shotaroikeda** [HackIllinois/api-2017#74]
We need a systematic way to handle events, for both the mobile apps and dashboard.

~~The specifics are highlighted [here](https://github.com/HackIllinois/api-2017/wiki/Events-Documentation).~~

I've already checked with @kpchi and @redsn0w422 to see if they are ok with the restrictions I've set in place. 

Since Vincent (no GitHub account) has to work under iOS limitations, there is some nuances about how to pick Event/Location names and descriptions. We need some training or some documentation highlighting limitations.

Limitations:
- Event Names should be under 25 characters. Overflow should end up in the description.
- Event Locations should be under 25 characters. Overflow should end up in the description.

**Comment by nmagerko** [comment]
Here is the proposed documentation. Moved from the Wiki

**GET /v1/events** <br />
Obtain current events.

Headers <br />
None

URL Parameters <br />
None

Response
```
{
	"meta": null,
	"data": {
		"id" : 1,
		"time": 1485488574,
		"last_updated": 1485488574,
		"locations": [
			{
				"id": 1,
				"name": "Thomas Siebel Center for Computer Science",
				"longitude": 40.113926,
				"latitude": -88.224916
			},
			{
				"id": 2,
				"name": "Electrical and Computer Engineering Building",
				"longitude": 40.114828,
				"latitude": -88.228049
			}
		],
		"location_id": 1,
		"name": "Breakfast",
		"qr_code": 1,
		"description": "Pancakes served for breakfast!"
	}
}
```

Some Notes on `location_id`:
Location IDs 1, 2, 3 are reserved for the following(?):

| Value | Corresponding Location |
| ----- | ---------------------- |
| 1     | Thomas Siebel Center for Computer Science |
| 2     | Electrical and Computer Engineering Building |
| 3     | Illini Union            |

A Location ID 4 or above indicates that the location is new.

Non-zero `qr_code` determines whether if we need a QR code for the event or not

---

**POST /v1/events/update** <br />
Update an event. The event to be updated will be based off the id.
The location WILL NOT be updated with this endpoint.

Headers <br />

| Header        | Description           | Required  |
| ------------- | --------------------- | --------- |
| `Authorization` | a valid authentication token from a SUPERUSER | Yes |

Request
```
{
	"data": {
		"id" : 1,
		"time": 1485488574,
		"name": "Brunch",
		"qr_code": 1,
		"description": "Sorry our breakfast came late! We will now serve cereal and pancakes!"
	}
}
```

Some notes:

`last_updated` will be based off of the current server time.
All parameters are required in order to update the row. Or the server will throw an error.


Response
```
{
	"meta": null,
}
```

Errors: <br />

| Error        | Source | Cause  |
| ------------ | ------ | ------ |
| InvalidHeaderError | `Authorization` | the authentication token was absent or invalid |
| InvalidParameterError | `Request Parameter(s)` | the request parameter did not contain all fields |

---

**POST /v1/events/update/location** <br />
Update a event's location. The location to be updated will be based off the id.
The location WILL NOT be updated with this endpoint.

Headers <br />

| Header        | Description           | Required  |
| ------------- | --------------------- | --------- |
| `Authorization` | a valid authentication token from a SUPERUSER | Yes |

Request
```
{
	"data": {
		[
			{
				"id": 1,
				"name": "Thomas Siebel Center for Computer Science",
				"longitude": 40.113926,
				"latitude": -88.224916
			},
			{
				"id": 2,
				"name": "Illini Union",
				"longitude": 40.113926,
				"latitude": -88.224916
			}
		]
	}
}
```

Response
```
{
	"meta": null,
}
```

Errors: <br>

| Error        | Source | Cause  |
| ------------ | ------ | ------ |
| InvalidHeaderError | `Authorization` | the authentication token was absent or invalid |
| InvalidParameterError | `Request Parameter(s)` | the request parameter did not contain all fields |

---

**Comment by nmagerko** [comment]
A few things:

Please don't put documentation for things in the Wiki when the feature does not exist. That doesn't make sense now that people are reading the documentation to develop on the front-end.

I'm not really sure that I understand what you mean about these limitations. I don't think the back-end should have to worry about character limits on a specific platform if we can/want to store more in a specific field. That's something the platform developer has to worry about.

Are you asking for this feature to be developed or is it a suggestion? I do not see this as being a very important feature when we know the events ahead of time, right?

**Comment by shotaroikeda** [comment]
Thanks Nick!
Unfortunately, it ended up being a pretty important feature because we actually don't know the events ahead of time. People in experience told me to work on this after consulting with KP and Yasha about our mobile application views.

If you don't mind, I'll be editing the stuff you have up top until the feature is finished!

**Comment by shotaroikeda** [comment]
The limitations are actually from mobile applications. There's no good way (it's an unsolved problem!) to shorten names to fit on the screen so that the user can still understand it. This why the restrictions are placed, and I plan to place them on the SQL database as well.

A good way to think about it is so that people come up with short, but concise names to represent the Event Name or Location.

**Comment by nmagerko** [comment]
@shotaroikeda I do not believe that's an unsolved problem. How is it possible that there's no way to make a GET request from the app and then do the exact transformation you talk about before displaying it?

Please show me something that explains this issue (image, blog from somewhere, etc). I just don't understand how an adapter for this information cannot be written, thereby imposing something on the data source (i.e. the API).

**Comment by nmagerko** [comment]
If what you're saying is that you want a "brief title" and "brief location" field in addition to the event name and event location fields, that seems reasonable. Otherwise I don't see a reason to put restrictions on fields that are easy for us to store.

**Comment by shotaroikeda** [comment]
Yep I do agree that we can potentially have that, but let me play devil's advocate here:
Why should we include those columns if we aren't going to use them*?

(Not sure if web is planning on using it, I'll make sure to find out on Thursday)

**Comment by tommypacker** [comment]
Wait, so do you want the actual title to be an abbreviated title?

**Comment by shotaroikeda** [comment]
The thing is, I don't want it to be abbreviated, but it should be "reasonably short". Personally, the best way we can guarantee that is by having a hard limit on a title.

Though that being said, we do lose out on flexibility here.

**Comment by shotaroikeda** [comment]
Maybe the best here is drawing a line as follows:
1. title+event name can be as long as they'd like
2. the "brief" title or event name should be limited to 25 characters.

**Comment by nmagerko** [comment]
Well up to 255 characters would probably be good for (1) but yeah. 

I don't see a reason not to store the full event name and description if it makes it easier for people looking at the data (i.e. staff members) to know what they're looking at, especially after the event is over. Storing 510 characters in addition to the 50 that we'd need already is negligible anyways.

I assume you're storing the start/end timestamps too right?

**Comment by shotaroikeda** [comment]
Ok sure. Yep, I forgot about adding the ending time stamp, so I will add that.

**Comment by nmagerko** [comment]
Implemented by #90
